bellman/src/constraint.rs

```diff
@@ -0,0 +1,131 @@
+use crate::{Namespace, domain, error};
+use domain::{Variable, LinearCombination, Index};
+use error::Result;
+
+use ff::ScalarEngine;
+
+/// Represents a constraint system which can have new variables
+/// allocated and constrains between them formed.
+pub trait ConstraintSystem<E>: Sized 
+where
+    E: ScalarEngine
+{
+    /// Represents the type of the "root" of this constraint system
+    /// so that nested namespaces can minimize indirection.
+    type Root: ConstraintSystem<E>;
+
+    /// Return the "one" input variable
+    fn one() -> Variable {
+        Variable::new_unchecked(Index::Input(0))
+    }
+
+    /// Allocate a private variable in the constraint system. The provided function is used to
+    /// determine the assignment of the variable. The given `annotation` function is invoked
+    /// in testing contexts in order to derive a unique name for this variable in the current
+    /// namespace.
+    fn alloc<F, A, AR>(&mut self, annotation: A, f: F) -> Result<Variable>
+    where
+        F: FnOnce() -> Result<E::Fr>,
+        A: FnOnce() -> AR,
+        AR: Into<String>;
+
+    /// Allocate a public variable in the constraint system. The provided function is used to
+    /// determine the assignment of the variable.
+    fn alloc_input<F, A, AR>(&mut self, annotation: A, f: F) -> Result<Variable>
+    where
+        F: FnOnce() -> Result<E::Fr>,
+        A: FnOnce() -> AR,
+        AR: Into<String>;
+
+    /// Enforce that `A` * `B` = `C`. The `annotation` function is invoked in testing contexts
+    /// in order to derive a unique name for the constraint in the current namespace.
+    fn enforce<A, AR, LA, LB, LC>(&mut self, annotation: A, a: LA, b: LB, c: LC)
```

Comment:
Don't be afraid to rename functions if there could be a more readable/understandable name. Nit: I would have put annotation last it makes more sense to put the more auxiliary parameters last. Also, a function should really be a verb rather than a noun. I'm not sure about the Rust specific practices but also if this could be called from 'self' rather than passed in a parameter it really should be.

Reply:
I’ve avoided breaking APIs with this first release and enforce is a major one used throughout the library. 

In terms of the name, enforce is also a verb and it is currently calling through self. 

This enforce method is probably the most frustrating to work with, and it won’t be until I’ve completely reworked the ‘gadget’ module that I’m confident in what to change. It involves passing a string (annotation) and three lambda arguments, some of which often do nothing  

Agreed on auxiliary parameters, unfortunately the naming of annotation is a bit weak. It’s actually more of a title for the piece of the circuit, so the enforce call would look something like this:

my_circuit.enforce(
   ‘xor gate’,
    .....
)

We can rename the parameter if you have any suggestions.